pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1152 @@
+	pub(crate) fn denote_collected_investment(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		collected: CollectedAmount<T::Balance>,
+	) -> DispatchResult {
+		// Increment by previously stored amounts (via `CollectedInvestmentHook`)
+		CollectedInvestment::<T>::mutate(who, investment_id, |collected_before| {
+			collected_before
+				.amount_collected
+				.ensure_add_assign(collected.amount_collected)?;
+			collected_before
+				.amount_payment
+				.ensure_add_assign(collected.amount_payment)?;
+			Ok::<(), DispatchError>(())
+		})?;
+
+		// Update invest state to decrease the unprocessed investing amount
+		let investing_amount = T::Investment::investment(who, investment_id)?;
+		let pre_state = InvestmentState::<T>::get(who, investment_id);
+		let post_state =
+			pre_state.transition(InvestTransition::CollectInvestment(investing_amount))?;
+		Self::apply_invest_state_transition(who, investment_id, post_state).map_err(|e| {
+			log::debug!("InvestState transition error: {:?}", e);
+			Error::<T>::from(InvestError::Collect)
+		})?;
+
+		Ok(())
+	}
+
+	/// Consumes the `CollectedInvestment` amounts and returns these.
+	///
+	/// NOTE: Converts the collected pool currency payment amount to foreign
+	/// currency via the `CurrencyConverter` trait.
+	pub(crate) fn transfer_collected_investment(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		foreign_payout_currency: T::CurrencyId,
+		pool_currency: T::CurrencyId,
+	) -> Result<ExecutedForeignCollectInvest<T::Balance>, DispatchError> {
+		let collected = CollectedInvestment::<T>::take(who, investment_id);
+		ensure!(
+			!collected.amount_payment.is_zero(),
+			Error::<T>::InvestError(InvestError::NothingCollected)
+		);
+
+		// Determine payout amount in foreign currency instead of current pool currency
+		// denomination
+		let amount_currency_payout = T::CurrencyConverter::stable_to_stable(
+			pool_currency,
+			collected.amount_payment,
+			foreign_payout_currency,
+		)?;
+
+		Ok(ExecutedForeignCollectInvest {
+			amount_currency_payout,
+			amount_tranche_tokens_payout: collected.amount_collected,
+		})
+	}
+
+	/// Increments the collected redemption amount and transitions redemption
+	/// state as a result of collecting the redemption.
+	///
+	/// NOTE: Neither initiates a swap from the collected pool currency into
+	/// foreign currency nor transfers back any currency to the investor. This
+	/// happens in `transfer_collected_redemption`.
+	pub(crate) fn denote_collected_redemption(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		collected: CollectedAmount<T::Balance>,
+	) -> DispatchResult {
+		let foreign_payout_currency =
+			RedemptionPayoutCurrency::<T>::get(who, investment_id).expect("Foreign redemption payout currency is set by initial increment before collecting is possible");
+		let pool_currency = T::PoolInspect::currency_for(investment_id.of_pool())
+			.expect("Impossible to collect redemption for non existing pool at this point");
```

Comment:
Same here I think. What if the user uses an erroneous `investment_id`? or the pool is removed?

Reply:
If the user uses an erroneous `investment_id`, then the `T::Investment::collect_redemption` call would already fail. Please note that `denote_collected_redemption` is only called through the `StatusNotificationHook` impl of `CollectedRedemptionHook` which is executed during `do_collect_redeem`.